p2p/src/utils/requester.rs

```diff
@@ -0,0 +1,387 @@
+    ///
+    /// If `shuffle` is true, the order of participants is shuffled before
+    /// a request is made. This is typically used when a request to the preferred
+    /// participant fails.
+    pub fn request(&mut self, shuffle: bool) -> Option<(PublicKey, ID)> {
+        // Prepare participant iterator
+        let participant_iter = if shuffle {
+            let mut participants = self.participants.iter().collect::<Vec<_>>();
+            participants.shuffle(&mut self.runtime);
+            Either::Left(participants.into_iter())
+        } else {
+            Either::Right(self.participants.iter())
+        };
+
+        // Look for a participant that can handle request
+        for (participant, _) in participant_iter {
+            // Check if me
+            if *participant == self.crypto.public_key() {
+                continue;
+            }
+
+            // Check if excluded
+            if self.excluded.contains(participant) {
+                continue;
+            }
+
+            // Check if rate limit is exceeded (and update rate limiter if not)
+            if self.rate_limiter.check_key(participant).is_err() {
+                continue;
+            }
+
+            // Compute ID
+            //
+            // As long as we don't have u64 requests outstanding, this is ok.
+            let id = self.id;
+            self.id = self.id.wrapping_add(1);
+
+            // Record request issuance time
+            let now = self.runtime.current();
+            self.requests.insert(id, (participant.clone(), now));
+            let deadline = now.checked_add(self.timeout).expect("time overflowed");
+            self.deadlines.put(id, deadline);
+            return Some((participant.clone(), id));
+        }
+        None
+    }
+
+    /// Resolve an outstanding request.
+    pub fn resolved(&mut self, id: ID) {
+        // Remove request
+        let Some((participant, start)) = self.cancel(id) else {
+            return;
+        };
+
+        // Get elapsed time
+        let Ok(elapsed) = self.runtime.current().duration_since(start) else {
+            return;
+        };
+        let elapsed = elapsed.as_millis();
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past.saturating_add(elapsed) / 2;
+        self.participants.put(participant, performance);
+    }
+
+    /// Timeout an outstanding request.
+    pub fn timeout(&mut self, id: ID) {
+        let Some((participant, _)) = self.cancel(id) else {
+            return;
+        };
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past.saturating_add(self.timeout.as_millis()) / 2;
+        self.participants.put(participant.clone(), performance);
+    }
```

Comment:
I wonder if these can actually just be combined into one function.

They are exactly the same in the case that the timeout happens exactly at the right time, and only slightly different otherwise. So would suggest removing the `timeout` function and having adding `let elapsed = min(elapsed, self.timeout).as_millis();`